[reland][BE][autograd Function] Raise an error if input is returned a…

soulitzer opened the pull request:
…s-is and saved for forward or backward in setup_context

Fixes #ISSUE_NUMBER

Relanding this in a new non-ghstack PR so I can import this to do co-dev

zou3519 reviewed (approved):
Good luck on the codev

atalman commented:
Please do not land it, we are trying to revert co-dev diff that is already landed for this PR

soulitzer committed:
[reland][BE][autograd Function] Raise an error if input is returned a… (#98051)

Summary:
…s-is and saved for forward or backward in setup_context

Fixes #ISSUE_NUMBER

Relanding this in a new non-ghstack PR so I can import this to do co-dev

Pull Request resolved: https://github.com/pytorch/pytorch/pull/98051

Reviewed By: zou3519

Differential Revision: D44747382

Pulled By: soulitzer

fbshipit-source-id: e3d7fad3c847d8ef96e2c06654ab9449771b092a
